osu.Game/Beatmaps/BeatmapManager.cs

```diff
@@ -112,29 +113,57 @@ public WorkingBeatmap CreateNew(RulesetInfo ruleset, APIUser user)
-            // fetch one of the existing difficulties to copy timing points and metadata from,
-            // so that the user doesn't have to fill all of that out again.
-            // this silently assumes that all difficulties have the same timing points and metadata,
-            // but cases where this isn't true seem rather rare / pathological.
-            var referenceBeatmap = GetWorkingBeatmap(beatmapSetInfo.Beatmaps.First());
+            var referenceBeatmap = creationParameters.ReferenceBeatmap;
+            var targetBeatmapSet = creationParameters.BeatmapSet;
 
-            var newBeatmapInfo = new BeatmapInfo(rulesetInfo, new BeatmapDifficulty(), referenceBeatmap.Metadata.DeepClone());
+            BeatmapInfo newBeatmapInfo;
+            IBeatmap newBeatmap;
+
+            if (creationParameters.CreateBlank)
+            {
+                newBeatmapInfo = new BeatmapInfo(creationParameters.Ruleset, new BeatmapDifficulty(), referenceBeatmap.Metadata.DeepClone());
+                newBeatmap = new Beatmap { BeatmapInfo = newBeatmapInfo };
+                foreach (var timingPoint in referenceBeatmap.ControlPointInfo.TimingPoints)
+                    newBeatmap.ControlPointInfo.Add(timingPoint.Time, timingPoint.DeepClone());
+            }
+            else
+            {
+                newBeatmap = referenceBeatmap.Clone();
+
+                var referenceBeatmapInfo = referenceBeatmap.BeatmapInfo;
+                newBeatmap.BeatmapInfo = newBeatmapInfo = new BeatmapInfo(referenceBeatmapInfo.Ruleset, referenceBeatmapInfo.Difficulty.Clone(), referenceBeatmapInfo.Metadata.DeepClone())
+                {
+                    // Only selected appropriate properties are copied over.
+                    // Things like database ID, online status/ID, MD5 hash, star rating, etc. are omitted
+                    // because they should not be copied over and/or they will be recomputed on save.
+                    AudioLeadIn = referenceBeatmapInfo.AudioLeadIn,
+                    StackLeniency = referenceBeatmapInfo.StackLeniency,
+                    SpecialStyle = referenceBeatmapInfo.SpecialStyle,
+                    LetterboxInBreaks = referenceBeatmapInfo.LetterboxInBreaks,
+                    WidescreenStoryboard = referenceBeatmapInfo.WidescreenStoryboard,
+                    EpilepsyWarning = referenceBeatmapInfo.EpilepsyWarning,
+                    SamplesMatchPlaybackRate = referenceBeatmapInfo.SamplesMatchPlaybackRate,
+                    DistanceSpacing = referenceBeatmapInfo.DistanceSpacing,
+                    BeatDivisor = referenceBeatmapInfo.BeatDivisor,
+                    GridSize = referenceBeatmapInfo.GridSize,
+                    TimelineZoom = referenceBeatmapInfo.TimelineZoom,
+                    Countdown = referenceBeatmapInfo.Countdown,
+                    CountdownOffset = referenceBeatmapInfo.CountdownOffset,
+                    Bookmarks = (int[])referenceBeatmapInfo.Bookmarks.Clone()
```

Comment:
Probably not a great sign that I can't see where the `DifficultyName` is being set to empty, but I don't think it should be.

It should be something like `Hard (Copy)`, both to identify the newly created difficulty better, and to avoid this kind of thing:

![osu! 2022-02-14 at 06 26 08](https://user-images.githubusercontent.com/191335/153811419-8c88e399-a9c9-4849-9ab5-5d46bfda8063.png)

Reproduction is saving twice without setting a difficulty name. Best way to avoid is to add ` (Copy)` each time until it succeeds.

Also I think after creating the difficulty it should force the user to the song setup screen to set a new difficulty name. Preferrably with the textbox pre-highlighted, but I'm not sure how simple that is to make happen.

Reply:
It's empty precisely because it isn't set after I made the copy process opt-in wrt the set of properties to copy. But given your requirement to append a suffix this will go away either way.

I'll see what I can do about navigating to metadata. Maybe that part should also apply to blank maps?

Reply:
> I'll see what I can do about navigating to metadata. Maybe that part should also apply to blank maps?

For sure.. I thought we were already doing that (are we not?)

https://github.com/ppy/osu/blob/a144d6f8d67b0fe6c90021b3309682f323da9322/osu.Game/Screens/Edit/Editor.cs#L248

Right now, the `artist` textbox is also automatically highlighted if empty, on arriving at the setup screen. Maybe this screen needs to accept a startup parameter of which field to select on load instead.

Reply:
I can see why the current logic would fail for new difficulties. The import/creation of the difficulty happens outside of the editor (i.e. at the editor loader level), so at the point of editor being re-entered the new difficulty is no longer "new".

Reply:
Probably fine to track as a follow-up task either way. Getting the difficulty name non-blank would suffice to get this one merged.

Reply:
The copy suffix will now apply - but this was substantially more painful to make happen than you would imagine, due to reference handling around cached working beatmaps again.

In short: In my initial diff I made the mistake of passing `playableBeatmap` into `switchToNewDifficulty()`. That instance comes from the working beatmap cache and doesn't have all the changes that are made inside the editor session. Now there is a flow to invalidate it, but it happens [when the editor is exited or suspended](https://github.com/ppy/osu/blob/fdddeb7ed26c5b80253f746e8a83b4ba5eee42ac/osu.Game/Screens/Edit/Editor.cs#L618-L632).

So to counteract that I was sort of forced to use `GetWorkingBeatmap()` in `EditorLoader` to ensure that the difficulty creation process has all the newest data. But I need to be extra careful to not touch any model instances that belong to that cached working beatmap, lest I pollute the state of the cache. So I end up [taking out both a fresh detached model instance and a working beatmap](https://github.com/ppy/osu/blob/e45a2ae0fc9b24af51454589be31923cb80543ca/osu.Game/Screens/Edit/EditorLoader.cs#L89-L91). The former is safe to modify to my heart's content without any fears, and the latter will be invalidated by the save flow if successful (and if it isn't successful - well, I'd hope that the save operation isn't changing anything in the beatmap).

I don't know. It's dodgy, but I don't know how to make work better.